File: 0332-reconstruct-itinerary/0332-reconstruct-itinerary.java
class Solution {
    public List<String> findItinerary(List<List<String>> tickets) {
        Map<String, PriorityQueue<String>> graph = new HashMap<>();

        for (List<String> ticket : tickets) {
            String departure = ticket.get(0);
            String arrival = ticket.get(1);

            graph.computeIfAbsent(departure, k -> new PriorityQueue<>()).add(arrival);
        }

        List<String> newItinerary = new ArrayList<>();
        Deque<String> stack = new ArrayDeque<>();
        stack.push("JFK");

        while (!stack.isEmpty()) {
            String currentAirport = stack.peek();

            if (graph.containsKey(currentAirport) && !graph.get(currentAirport).isEmpty()) {
                stack.push(graph.get(currentAirport).poll());
            } else {
                newItinerary.add(stack.pop());
            }
        }

        Collections.reverse(newItinerary);
        return newItinerary;        
    }
}

File: 0332-reconstruct-itinerary/0332-reconstruct-itinerary.py
class Solution(object):
    def findItinerary(self, tickets):
        """
        :type tickets: List[List[str]]
        :rtype: List[str]
        """
        adj = defaultdict(list)

        for fr, to in tickets:
            heapq.heappush(adj[fr], to)

        path = []
        def dfs(curr):
            while adj[curr]:
                dfs(heapq.heappop(adj[curr]))
            path.append(curr)
        
        dfs("JFK")
        return path[::-1]